mne/time_frequency/spectrum.py

```diff
@@ -1179,34 +1214,73 @@ def __getitem__(self, item):
         return BaseRaw._getitem(self, item, return_times=False)
 
 
-def _check_data_shape(data, freqs, info, ndim):
-    if data.ndim != ndim:
-        raise ValueError(f"Data must be a {ndim}D array.")
+def _check_data_shape(
+    data, info, freqs, is_epoched, events, method, standard_ndim, extra_ndim, mt_weights
+):
+    _check_option("data.ndim", data.ndim, (standard_ndim, extra_ndim))
+
+    dims = ()
+    if is_epoched:
+        dims += ("epoch",)
+        if events is not None and data.shape[0] != events.shape[0]:
+            raise ValueError(
+                f"The first dimension of `data` ({data.shape[0]}) must match the first "
+                f"dimension of `events` ({events.shape[0]})."
+            )
+
+    dims += ("channel",)
     want_n_chan = _pick_data_channels(info).size
-    want_n_freq = freqs.size
-    got_n_chan, got_n_freq = data.shape[-2:]
+    got_n_chan = data.shape[list(dims).index("channel")]
     if got_n_chan != want_n_chan:
         raise ValueError(
-            f"The number of channels in `data` ({got_n_chan}) must match the "
-            f"number of good data channels in `info` ({want_n_chan})."
+            f"The number of channels in `data` ({got_n_chan}) must match the number of "
+            f"good data channels in `info` ({want_n_chan})."
         )
+
+    if data.ndim == extra_ndim:  # i.e. segments or tapers present
+        _check_option(
+            "method", method, ("multitaper", "welch"), f" when data.ndim={extra_ndim}"
+        )  # require method specified to differentiate segments from tapers
+        if method == "multitaper":
+            actual = None if mt_weights is None else mt_weights.size
+            expected = data.shape[-2]
+            if actual != expected:
+                raise ValueError(
+                    f"Expected size of `mt_weights` to be {expected}, got {actual}."
+                )
+            dims += ("taper", "freq")
+        else:  # i.e. welch
+            dims += ("freq", "segment")
+    else:
+        dims += ("freq",)
```

Comment:
One of the hang-ups in #11803 seemed to be the question of differentiating complex multitaper `(chans x tapers x freqs)` data from unaggregated Welch `(chans x freqs x segments)` data in `(Epochs)SpectrumArray` classes.

Here I took from @alexrockhill's suggestion that if data has this extra segments/tapers dimension, the user is required to specify the method (rather than it being forced to be `"unknown"`) so that the dimensions can be discerned.

Reply:
Sure, I get you.

> you could do `dimnames=None` and fill in the default of `("epochs", "channels", "freqs")` in the code... but I sorta like having the "default" be present in the param defaults, as it makes it more obvious how to alter it if needed.

Yeah, I prefer that as well.

I've given this a go. There are some extra checks now since there are additional considerations to make when allowing the user to pass in the dims. This is not so much in terms of actually packaging the data in the object, but for later calls to plotting functions, conversion to dataframe, etc... Still, some checks could probably be streamlined.

https://github.com/mne-tools/mne-python/blob/322ae92f4879d24a31d9ddbd1c504c594dbd1966/mne/time_frequency/spectrum.py#L1217-L1271

Some explanation:
- 1224-29 - need to enforce a set of recognised dimensions
- 1230-31 - at the very least, there needs to be a chan & freq dimension
- 1233-44 - like before, check n_chans matches that in data, but also check that chans is in correct place
- 1246-62 - for plotting & to_df calls, there are assumptions that if data contains tapers, it does not have segments (& vice versa), and with earlier checks (also in `(Epochs)SpectrumArray.__init__()`), we are enforcing that there can be only one of these additional dimensions
- 1249-52 - generally in `BaseSpectrum` the code is pretty flexible for the actual axis index of each dimension, but `_psd_from_mt` (called when plotting to aggr over tapers and convert to power) has index of tapers in penultimate axis fixed
- 1264-71 - like before, check n_freqs matches that in freqs, but don't need to bother checking whether freqs in some weird position since above checks rule that out
- 1253-60 - still make the assumption that if tapers are present, they should have weights